core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1180,26 +1194,36 @@ private int reuseOrCreateNewSortOrderId(SortOrder newOrder) {
       return newOrderId;
     }
 
-    private void setCurrentSnapshot(Snapshot snapshot, Long currentTimestampMillis) {
-      if (snapshot == null) {
-        this.currentSnapshotId = -1;
-        snapshotLog.clear();
-        changes.add(new MetadataUpdate.SetCurrentSnapshot(null));
-        return;
-      }
-
-      if (currentSnapshotId == snapshot.snapshotId()) {
-        return;
+    private void setBranchSnapshot(Snapshot snapshot, String branch, Long currentTimestampMillis) {
+      long replacementSnapshotId = snapshot.snapshotId();
+      SnapshotRef ref = refs.get(branch);
+      if (ref != null) {
+        ValidationException.check(ref.isBranch(), "Cannot update branch: %s is a tag", branch);
+        if (ref.snapshotId() == replacementSnapshotId) {
+          return;
+        }
       }
 
       ValidationException.check(formatVersion == 1 || snapshot.sequenceNumber() <= lastSequenceNumber,
           "Last sequence number %s is less than existing snapshot sequence number %s",
           lastSequenceNumber, snapshot.sequenceNumber());
 
       this.lastUpdatedMillis = currentTimestampMillis != null ? currentTimestampMillis : snapshot.timestampMillis();
-      this.currentSnapshotId = snapshot.snapshotId();
-      snapshotLog.add(new SnapshotLogEntry(lastUpdatedMillis, snapshot.snapshotId()));
-      changes.add(new MetadataUpdate.SetCurrentSnapshot(snapshot.snapshotId()));
+
+      if (SnapshotRef.MAIN_BRANCH.equals(branch)) {
```

Comment:
looks like the check for main branch is done in many places, should we also have a method for `ref.isMainBranch()`?

Reply:
I think the problem is that there aren't many places where we actually have a `SnapshotRef` instance. Here, for example, we don't necessarily have one because `ref` might be null. So we actually want to use the string branch name.

Below, we have a `SetSnapshotRef` instance, which is a change. And in places like `removeBranch`, we do the check before loading the ref. If there is no main ref to remove, we should still ensure that we set current-snapshot-id appropriately.